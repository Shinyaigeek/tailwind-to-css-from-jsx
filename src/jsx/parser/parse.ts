import { babelParse } from "@sxzz/ast-kit";
import { createErr, createOk, Result } from "option-t/plain_result";

export const parseJSX: (
  code: string,
  isTSX: boolean,
) => Result<ReturnType<typeof babelParse>, Error> = function (code, isTSX) {
  try {
    return createOk(babelParse(code, isTSX ? "tsx" : "jsx"));
  } catch (e) {
    return createErr(
      new Error("Can't parse passed file", {
        cause: e,
      }),
    );
  }
};
