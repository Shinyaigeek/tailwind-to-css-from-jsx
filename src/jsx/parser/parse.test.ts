import { assertEquals } from "jsr:@std/assert";
import { createAssertSnapshot } from "jsr:@std/testing/snapshot";
import { stripAnsiCode } from "jsr:@std/fmt/colors";
import { parseJSX } from "./parse.ts";
import { isOk, unwrapOk } from "option-t/plain_result";

const assertSnapshot = createAssertSnapshot({
  dir: ".snaps",
});

const assertMonochromeSnapshot = createAssertSnapshot<string>(
  { serializer: stripAnsiCode },
  assertSnapshot,
);

Deno.test({
  name: "parse should return an AST on JSX",
  fn: async function (t) {
    const code = `function Test() {
            return <div></div>
        }`;
    const res = parseJSX(code, false);

    assertEquals(isOk(res), true);
    await assertMonochromeSnapshot(t, JSON.stringify(unwrapOk(res)));
  },
});

Deno.test({
  name: "parse should return an AST on TSX",
  fn: async function (t) {
    const code = `function Test(hello: string) {
            return <div>{hello}</div>
        }`;
    const res = parseJSX(code, true);

    assertEquals(isOk(res), true);
    await assertMonochromeSnapshot(t, JSON.stringify(unwrapOk(res)));
  },
});

Deno.test({
  name: "parse should return an error on invalid code",
  fn: function () {
    const code = `function Test(hello: string) {
            return <div>{hello}</div>
        `;
    const res = parseJSX(code, true);

    assertEquals(isOk(res), false);
  },
});
