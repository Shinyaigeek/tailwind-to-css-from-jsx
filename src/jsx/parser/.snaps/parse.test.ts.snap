export const snapshot = {};

snapshot[`parse should return an AST on JSX 1`] = `{"type":"Program","start":0,"end":58,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":3,"column":9,"index":58}},"sourceType":"module","interpreter":null,"body":[{"type":"FunctionDeclaration","start":0,"end":58,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":3,"column":9,"index":58}},"id":{"type":"Identifier","start":9,"end":13,"loc":{"start":{"line":1,"column":9,"index":9},"end":{"line":1,"column":13,"index":13},"identifierName":"Test"},"name":"Test"},"generator":false,"async":false,"params":[],"body":{"type":"BlockStatement","start":16,"end":58,"loc":{"start":{"line":1,"column":16,"index":16},"end":{"line":3,"column":9,"index":58}},"body":[{"type":"ReturnStatement","start":30,"end":48,"loc":{"start":{"line":2,"column":12,"index":30},"end":{"line":2,"column":30,"index":48}},"argument":{"type":"JSXElement","start":37,"end":48,"loc":{"start":{"line":2,"column":19,"index":37},"end":{"line":2,"column":30,"index":48}},"openingElement":{"type":"JSXOpeningElement","start":37,"end":42,"loc":{"start":{"line":2,"column":19,"index":37},"end":{"line":2,"column":24,"index":42}},"name":{"type":"JSXIdentifier","start":38,"end":41,"loc":{"start":{"line":2,"column":20,"index":38},"end":{"line":2,"column":23,"index":41}},"name":"div"},"attributes":[],"selfClosing":false},"closingElement":{"type":"JSXClosingElement","start":42,"end":48,"loc":{"start":{"line":2,"column":24,"index":42},"end":{"line":2,"column":30,"index":48}},"name":{"type":"JSXIdentifier","start":44,"end":47,"loc":{"start":{"line":2,"column":26,"index":44},"end":{"line":2,"column":29,"index":47}},"name":"div"}},"children":[]}}],"directives":[]}}],"directives":[],"extra":{"topLevelAwait":false},"errors":[],"comments":[]}`;

snapshot[`parse should return an AST on TSX 1`] = `{"type":"Program","start":0,"end":78,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":3,"column":9,"index":78}},"sourceType":"module","interpreter":null,"body":[{"type":"FunctionDeclaration","start":0,"end":78,"loc":{"start":{"line":1,"column":0,"index":0},"end":{"line":3,"column":9,"index":78}},"id":{"type":"Identifier","start":9,"end":13,"loc":{"start":{"line":1,"column":9,"index":9},"end":{"line":1,"column":13,"index":13},"identifierName":"Test"},"name":"Test"},"generator":false,"async":false,"params":[{"type":"Identifier","start":14,"end":27,"loc":{"start":{"line":1,"column":14,"index":14},"end":{"line":1,"column":27,"index":27},"identifierName":"hello"},"name":"hello","typeAnnotation":{"type":"TSTypeAnnotation","start":19,"end":27,"loc":{"start":{"line":1,"column":19,"index":19},"end":{"line":1,"column":27,"index":27}},"typeAnnotation":{"type":"TSStringKeyword","start":21,"end":27,"loc":{"start":{"line":1,"column":21,"index":21},"end":{"line":1,"column":27,"index":27}}}}}],"body":{"type":"BlockStatement","start":29,"end":78,"loc":{"start":{"line":1,"column":29,"index":29},"end":{"line":3,"column":9,"index":78}},"body":[{"type":"ReturnStatement","start":43,"end":68,"loc":{"start":{"line":2,"column":12,"index":43},"end":{"line":2,"column":37,"index":68}},"argument":{"type":"JSXElement","start":50,"end":68,"loc":{"start":{"line":2,"column":19,"index":50},"end":{"line":2,"column":37,"index":68}},"openingElement":{"type":"JSXOpeningElement","start":50,"end":55,"loc":{"start":{"line":2,"column":19,"index":50},"end":{"line":2,"column":24,"index":55}},"name":{"type":"JSXIdentifier","start":51,"end":54,"loc":{"start":{"line":2,"column":20,"index":51},"end":{"line":2,"column":23,"index":54}},"name":"div"},"attributes":[],"selfClosing":false},"closingElement":{"type":"JSXClosingElement","start":62,"end":68,"loc":{"start":{"line":2,"column":31,"index":62},"end":{"line":2,"column":37,"index":68}},"name":{"type":"JSXIdentifier","start":64,"end":67,"loc":{"start":{"line":2,"column":33,"index":64},"end":{"line":2,"column":36,"index":67}},"name":"div"}},"children":[{"type":"JSXExpressionContainer","start":55,"end":62,"loc":{"start":{"line":2,"column":24,"index":55},"end":{"line":2,"column":31,"index":62}},"expression":{"type":"Identifier","start":56,"end":61,"loc":{"start":{"line":2,"column":25,"index":56},"end":{"line":2,"column":30,"index":61},"identifierName":"hello"},"name":"hello"}}]}}],"directives":[]}}],"directives":[],"extra":{"topLevelAwait":false},"errors":[],"comments":[]}`;
